File: Julia/bert-fine-tuning.jl
mutable struct MockBertEncoder
    hidden_size::Int
    num_layers::Int
    layers::Vector
    layer_frozen::Vector{Bool}
end

function MockBertEncoder(hidden_size::Int=768, num_layers::Int=12)
    layers = [randn(hidden_size, hidden_size) .* 0.01 for _ in 1:num_layers]
    layer_frozen = fill(false, num_layers)
    MockBertEncoder(hidden_size, num_layers, layers, layer_frozen)
end

function freeze_layers!(encoder::MockBertEncoder, layer_indices)
    for idx in layer_indices
        if 1 <= idx <= encoder.num_layers
            encoder.layer_frozen[idx] = true
        end
    end
end

function unfreeze_all!(encoder::MockBertEncoder)
    encoder.layer_frozen .= false
end

function forward(encoder::MockBertEncoder, embeddings)
    x = embeddings
    for layer in encoder.layers
        x = x * layer .+ x
    end
    x
end

mutable struct BertForSequenceClassification
    encoder::MockBertEncoder
    classifier
    bias
    freeze_bert::Bool
end

function BertForSequenceClassification(hidden_size::Int, num_labels::Int; freeze_bert::Bool=false)
    encoder = MockBertEncoder(hidden_size)
    classifier = randn(hidden_size, num_labels) .* 0.02
    bias = zeros(num_labels)
    model = BertForSequenceClassification(encoder, classifier, bias, freeze_bert)
    if freeze_bert
        freeze_layers!(model.encoder, 1:12)
    end
    model
end

function forward(model::BertForSequenceClassification, embeddings)
    hidden_states = forward(model.encoder, embeddings)
    cls_representation = hidden_states[:, 1, :]
    cls_representation * model.classifier .+ model.bias
end

mutable struct BertForTokenClassification
    encoder::MockBertEncoder
    classifier
    bias
end

function BertForTokenClassification(hidden_size::Int, num_labels::Int)
    encoder = MockBertEncoder(hidden_size)
    classifier = randn(hidden_size, num_labels) .* 0.02
    bias = zeros(num_labels)
    BertForTokenClassification(encoder, classifier, bias)
end

function forward(model::BertForTokenClassification, embeddings)
    hidden_states = forward(model.encoder, embeddings)
    hidden_states * model.classifier .+ model.bias
end

File: R/bert-fine-tuning.R
MockBertEncoder <- setRefClass(
  "MockBertEncoder",
  fields = list(
    hidden_size = "numeric",
    num_layers = "numeric",
    layers = "list",
    layer_frozen = "logical"
  ),
  methods = list(
    initialize = function(hidden_size = 768, num_layers = 12) {
      hidden_size <<- hidden_size
      num_layers <<- num_layers
      layers <<- lapply(seq_len(num_layers), function(i) matrix(rnorm(hidden_size * hidden_size, sd = 0.01), nrow = hidden_size, ncol = hidden_size))
      layer_frozen <<- rep(FALSE, num_layers)
    },
    freeze_layers = function(layer_indices) {
      for (idx in layer_indices) {
        if (idx >= 1 && idx <= num_layers) {
          layer_frozen[idx] <<- TRUE
        }
      }
    },
    unfreeze_all = function() {
      layer_frozen <<- rep(FALSE, num_layers)
    },
    forward = function(embeddings) {
      x <- embeddings
      for (layer in layers) {
        x <- x %*% layer + x
      }
      x
    }
  )
)

BertForSequenceClassification <- setRefClass(
  "BertForSequenceClassification",
  fields = list(
    encoder = "MockBertEncoder",
    classifier = "matrix",
    bias = "numeric",
    freeze_bert = "logical"
  ),
  methods = list(
    initialize = function(hidden_size, num_labels, freeze_bert = FALSE) {
      encoder <<- MockBertEncoder$new(hidden_size)
      classifier <<- matrix(rnorm(hidden_size * num_labels, sd = 0.02), nrow = hidden_size, ncol = num_labels)
      bias <<- numeric(num_labels)
      freeze_bert <<- freeze_bert
      if (freeze_bert) {
        encoder$freeze_layers(1:12)
      }
    },
    forward = function(embeddings) {
      hidden_states <- encoder$forward(embeddings)
      cls_representation <- hidden_states[, 1, ]
      cls_representation %*% classifier + bias
    }
  )
)

BertForTokenClassification <- setRefClass(
  "BertForTokenClassification",
  fields = list(
    encoder = "MockBertEncoder",
    classifier = "matrix",
    bias = "numeric"
  ),
  methods = list(
    initialize = function(hidden_size, num_labels) {
      encoder <<- MockBertEncoder$new(hidden_size)
      classifier <<- matrix(rnorm(hidden_size * num_labels, sd = 0.02), nrow = hidden_size, ncol = num_labels)
      bias <<- numeric(num_labels)
    },
    forward = function(embeddings) {
      hidden_states <- encoder$forward(embeddings)
      hidden_states %*% classifier + bias
    }
  )
)
